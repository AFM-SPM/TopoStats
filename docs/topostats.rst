topostats package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   topostats.logs
   topostats.tracing

Submodules
----------

.. toctree::
   :maxdepth: 4

   topostats.entry_point
   topostats.filters
   topostats.grains
   topostats.grainstats
   topostats.io
   topostats.plotting
   topostats.plottingfuncs
   topostats.processing
   topostats.run_topostats
   topostats.scars
   topostats.statistics
   topostats.theme
   topostats.thresholds
   topostats.utils
   topostats.validation

Module contents
---------------

.. automodule:: topostats
   :members:
   :undoc-members:
   :show-inheritance:
