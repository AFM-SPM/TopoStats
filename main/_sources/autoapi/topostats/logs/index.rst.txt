topostats.logs
==============

.. py:module:: topostats.logs

.. autoapi-nested-parse::

   
   Initialise login module.
















   ..
       !! processed by numpydoc !!


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/topostats/logs/logs/index


