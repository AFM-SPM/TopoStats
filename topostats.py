"""
File for tracking hard-coded variables (for now)
"""

import argparse

### List of hard-coded variables in pywgwytracing.py

# global variables
gwy_colobar_ln48 = "/module/pixmap/ztype"
gwy_linematchmethod_ln51= '/module/linematch/method'
gwy_linematchmaxdegree_ln52 = "/module/linematch/max_degree"
gwy_polylevel_coldegree_ln53 = "/module/polylevel/col_degree"

# heightediting()
datafield_grains_mark_height_ln128 = int()
gwy_polylevel_masking_ln131 = int()
gwy_linematch_masking_ln135 = int()
filterwidth_ln144 = float()

# editfile()
datafield_markheight_ln170 = int()
gwy_polylevel_masking_ln173 = int()
gwy_linematch_masking_ln177 = gwy_linematch_masking_ln135
gwy_polylevel_masking_ln181 = int()
datafield_gaussfilter_ln195 = int()

# grainfinding()
grain_gauss_ln211 = float()
datafield_mask_outliers_ln220 = float()

# remove_large_objects()
datafield_mask_outliers_ln257 = float()

# remove_small_objects()
datafield_mask_outliers_ln283 = float()

# grain_thinning()
gaussian_size_ln488 = float()

# savefiles()
savefile_type_ln535 = ["/module/pixmap/xytype"]

# saveunknownfiles()
savefile_type_ln573 = ["/module/pixmap/xytype"]

# savecroppedfiles()
savefile_type_ln645 = ["/module/pixmap/xytype"]

# __main__()
path_ln709 = str()
fileend_ln712 = []
filetype_ln713 = str()
extension_ln715 = str()
minheightscale_ln717 = float()
maxheightscale_ln718 = float()
minarea_ln722 = float()
maxdeviation_ln726 = float()
mindeviation_ln727 = float()
cropwidth_ln733 = float()
splitwidth_ln734 = float()
bins_ln736 = int()
channels_ln755 = list()


### dnatracing.py

# getNumpyArraysfromGwyddion()
sigma_ln94 = float()

# getDisorderedTrace()
sigma_ln109 = float()

# purgeObviousCrap()
maxlen_ln126 = int()

# getFittedTraces()
index_width_ln212 = int()

# getSplinedTraces()
step_size_ln376 = float()
interp_size_ln377 = float()

### tracingfuncs.py

# I didn't find any hard-coded variables so far

### statsplotting.py
### TFOplot.py
### traceplotting.py

# not checking these for now due to ongoing merging project
