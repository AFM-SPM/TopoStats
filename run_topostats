#!/usr/bin/env python3
"""Topotracing"""
import argparse as arg
from asyncio.log import logger
from collections import defaultdict

# from email.policy import default
from functools import partial
from multiprocessing import Pool
from pathlib import Path
from typing import Union, Dict

import pandas as pd
import numpy as np
from tqdm import tqdm
import numpy as np

from topostats.filters import Filters
from topostats.grains import Grains
from topostats.grainstats import GrainStats, get_grainstats
from topostats.io import read_yaml
from topostats.logs.logs import setup_logger, LOGGER_NAME
from topostats.plottingfuncs import plot_and_save
from topostats.thresholds import threshold
from topostats.tracing.dnatracing import dnaTrace
from topostats.utils import convert_path, find_images, update_config

LOGGER = setup_logger(LOGGER_NAME)


def create_parser() -> arg.ArgumentParser:
    """Create a parser for reading options."""
    parser = arg.ArgumentParser(
        description="Process AFM images. Additional arguments over-ride those in the configuration file."
    )
    parser.add_argument(
        "-c",
        "--config_file",
        dest="config_file",
        required=True,
        help="Path to a YAML configuration file.",
    )
    parser.add_argument(
        "-b",
        "--base_dir",
        dest="base_dir",
        type=str,
        required=False,
        help="Base directory to scan for images.",
    )
    parser.add_argument(
        "-f",
        "--file_ext",
        dest="file_ext",
        type=str,
        required=False,
        help="File extension to scan for.",
    )
    parser.add_argument(
        "--channel",
        dest="channel",
        type=str,
        required=False,
        help="Channel to extract.",
    )
    parser.add_argument(
        "-o",
        "--output_dir",
        dest="output_dir",
        type=str,
        required=False,
        help="Output directory to write results to.",
    )
    parser.add_argument(
        "-s",
        "--save_plots",
        dest="save_plots",
        type=bool,
        required=False,
        help="Whether to save plots.",
    )
    parser.add_argument(
        "-a",
        "--amplify_level",
        dest="amplify_level",
        type=float,
        required=False,
        help="Amplify signals by the given factor.",
    )
    parser.add_argument(
        "-t",
        "--threshold_method",
        dest="threshold_method",
        required=False,
        help="Method used for thresholding.",
    )
    parser.add_argument(
        "--otsu_threshold_multiplier",
        dest="otsu_threshold_multiplier",
        required=False,
        help="Factor to scale threshold during grain finding.",
    )
    parser.add_argument("-m", "--mask", dest="mask", type=bool, required=False, help="Mask the image.")
    parser.add_argument(
        "-q",
        "--quiet",
        dest="quiet",
        type=bool,
        required=False,
        help="Toggle verbosity.",
    )
    parser.add_argument(
        "-w",
        "--warnings",
        dest="quiet",
        type=str,
        required=False,
        help="Whether to ignore warnings.",
    )
    return parser


def process_scan(
    image_path: Union[str, Path] = None,
    channel: str = "Height",
    amplify_level: float = 1.0,
    filter_threshold_method: str = "otsu",
    filter_otsu_threshold_multiplier: Union[int, float] = 1.7,
    filter_threshold_std_dev=1.0,
    filter_threshold_abs_lower=None,
    filter_threshold_abs_upper=None,
    gaussian_size: Union[int, float] = 2,
    gaussian_mode: str = "nearest",
    absolute_smallest_grain_size=None,
    background: float = 0.0,
    grains_threshold_method: str = "otsu",
    grains_otsu_threshold_multiplier: Union[int, float] = 1.7,
    grains_threshold_std_dev=1.0,
    grains_threshold_abs_lower=None,
    grains_threshold_abs_upper=None,
    save_plots: bool = True,
    output_dir: Union[str, Path] = "output",
) -> None:
    """Process a single image, filtering, finding grains and calculating their statistics.

    Parameters
    ----------
    image_path : Union[str, Path]
        Path to image to process.
    channel : str
        Channel to extract and process, default 'height'.
    amplify_level : float
        Level to amplify image prior to processing by.
    threshold_method: str
        Method for determining threshold to mask values, default is 'otsu'.
    gaussian_size : Union[int, float]
        Minimum grain size in nanometers (nm).
    mode : str
        Mode for filtering (default is 'nearest').
    otsu_threshold_multiplier : Union[int, float]
        Factor by which lower threshold is to be scaled prior to masking.
    background : float
    save_plots : bool
        Flag as to whether to save plots to PNG files.
    output_dir : Union[str, Path]
        Path to output directory for saving results.

    Examples
    --------

    from topostats.topotracing import process_scan

    process_scan(image_path='minicircle.spm',
                 save_plots=True,
                 output_dir='output/')

    """
    LOGGER.info(f"Processing : {image_path}")
    # Filter Image :
    #
    # The Filters class has a convenience method that runs the instantiated class in full.
    filtered_image = Filters(
        image_path,
        threshold_method=filter_threshold_method,
        threshold_std_dev=filter_threshold_std_dev,
        threshold_absolute_lower=filter_threshold_abs_lower,
        threshold_absolute_upper=filter_threshold_abs_upper,
        channel=channel,
        amplify_level=amplify_level,
        output_dir=output_dir / image_path.stem / "filters",
    )
    filtered_image.filter_image()

    # Find Grains :
    #
    # The Grains class also has a convenience method that runs the instantiated class in full.
    grains = Grains(
        image=filtered_image.images["zero_averaged_background"],
        filename=filtered_image.filename,
        pixel_to_nm_scaling=filtered_image.pixel_to_nm_scaling,
        gaussian_size=gaussian_size,
        gaussian_mode=gaussian_mode,
        threshold_method=grains_threshold_method,
        otsu_threshold_multiplier=grains_otsu_threshold_multiplier,
        threshold_std_dev=grains_threshold_std_dev,
        threshold_absolute_lower=grains_threshold_abs_lower,
        threshold_absolute_upper=grains_threshold_abs_upper,
        absolute_smallest_grain_size=absolute_smallest_grain_size,
        background=background,
        base_output_dir=output_dir / filtered_image.filename / "grains",
    )
    grains.find_grains()

    # Grainstats :
    #
    # Ditto for the GrainStats class.
    grainstats = GrainStats(
        data=grains.images["gaussian_filtered"],
        labelled_data=grains.images["labelled_regions"],
        pixel_to_nanometre_scaling=filtered_image.pixel_to_nm_scaling,
        img_name=filtered_image.filename,
        output_dir=output_dir,
    )
    grain_statistics = grainstats.calculate_stats()

    # Run dnatracing
    # np.savetxt(Path("/home/neil/tmp/TopoStats/TOPOSTATS_PIXELS.csv"), filtered_image.images["pixels"], delimiter=",")
    # np.savetxt(Path("/home/neil/tmp/TopoStats/TOPOSTATS_GRAINS.csv"), grains.images["labelled_regions"], delimiter=",")
    # np.savetxt(
    #     Path("/home/neil/tmp/TopoStats/TOPOSTATS_GAUSSIAN_FILTERED.csv"),
    #     grains.images["gaussian_filtered"],
    #     delimiter=",",
    # )
    print(f"FROM run_topostats : {grains.images['labelled_regions']}")
    #    print(f"FROM run_topostats : {grains.images['labelled_regions'].tolist()}")
    dna_traces = dnaTrace(
        full_image_data=grains.images["gaussian_filtered"].T,
        grains=grains.images["labelled_regions"],
        afm_image_name=filtered_image.filename,
        pixel_size=filtered_image.pixel_to_nm_scaling,
        number_of_columns=grains.images["labelled_regions"].shape[0],
        number_of_rows=grains.images["labelled_regions"].shape[1],
    )

    # Optionally plot all stages
    if save_plots:
        # Filtering stage
        for plot_name, array in filtered_image.images.items():
            if plot_name not in ["scan_raw", "extracted_channel"]:
                PLOT_DICT[plot_name]["output_dir"] = Path(output_dir) / filtered_image.filename
                plot_and_save(array, **PLOT_DICT[plot_name])
        # Grain stage - only if we have grains
        if len(grains.region_properties) > 0:
            for plot_name, array in grains.images.items():
                PLOT_DICT[plot_name]["output_dir"] = Path(output_dir) / filtered_image.filename
                plot_and_save(array, **PLOT_DICT[plot_name])
            # Make a plot of coloured regions with bounding boxes
            plot_and_save(
                grains.images["coloured_regions"],
                Path(output_dir) / filtered_image.filename,
                **PLOT_DICT["bounding_boxes"],
                region_properties=grains.region_properties,
            )
            plot_and_save(
                grains.images["labelled_regions"],
                Path(output_dir) / filtered_image.filename,
                **PLOT_DICT["coloured_boxes"],
                region_properties=grains.region_properties,
            )
        # Grainstats
        # FIXME : Include this


def main():
    """Run processing."""

    # Parse command line options, load config and update with command line options
    parser = create_parser()
    args = parser.parse_args()
    config = read_yaml(args.config_file)
    config = update_config(config, args)

    LOGGER.info(f"Configuration file loaded from    : {args.config_file}")
    LOGGER.info(f'Scanning for images in            : {config["base_dir"]}')
    LOGGER.info(f'Output directory                  : {config["output_dir"]}')
    LOGGER.info(f'Looking for images with extension : {config["file_ext"]}')
    img_files = find_images(config["base_dir"])
    LOGGER.info(f'Images with extension {config["file_ext"]} in {config["base_dir"]} : {len(img_files)}')

    if config["quiet"]:
        LOGGER.setLevel("ERROR")
    # For debugging (as Pool makes it hard to find things when they go wrong)
    # for x in img_files:
    #     process_scan(
    #         image_path=x,
    #         channel=config["channel"],
    #         amplify_level=config["amplify_level"],
    #         threshold_method=config["threshold_method"],
    #         otsu_threshold_multiplier=config["grains"]["otsu_threshold_multiplier"],
    #         gaussian_size=config["grains"]["gaussian_size"],
    #         gaussian_mode=config["grains"]["gaussian_mode"],
    #         background=config["grains"]["background"],
    #         output_dir=config["output_dir"],
    #     )
    processing_function = partial(
        process_scan,
        channel=config["channel"],
        amplify_level=config["amplify_level"],
        filter_threshold_method=config["filter"]["threshold"]["method"],
        filter_otsu_threshold_multiplier=config["filter"]["threshold"]["otsu_multiplier"],
        filter_threshold_std_dev=config["filter"]["threshold"]["std_dev"],
        filter_threshold_abs_lower=config["filter"]["threshold"]["absolute"][0],
        filter_threshold_abs_upper=config["filter"]["threshold"]["absolute"][1],
        gaussian_size=config["grains"]["gaussian_size"],
        gaussian_mode=config["grains"]["gaussian_mode"],
        absolute_smallest_grain_size=config["grains"]["absolute_smallest_grain_size"],
        background=config["grains"]["background"],
        output_dir=config["output_dir"],
        grains_threshold_method=config["grains"]["threshold"]["method"],
        grains_otsu_threshold_multiplier=config["grains"]["threshold"]["otsu_multiplier"],
        grains_threshold_std_dev=config["grains"]["threshold"]["std_dev"],
        grains_threshold_abs_lower=config["grains"]["threshold"]["absolute"][0],
        grains_threshold_abs_upper=config["grains"]["threshold"]["absolute"][1],
    )

    with Pool(processes=config["cores"]) as pool:
        with tqdm(
            total=len(img_files),
            desc=f'Processing images from {config["base_dir"]}, results are under {config["output_dir"]}',
        ) as pbar:
            for _ in pool.imap_unordered(processing_function, img_files):
                pbar.update()


if __name__ == "__main__":
    main()
