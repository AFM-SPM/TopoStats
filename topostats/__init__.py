"""Topostats"""
from importlib.metadata import version

from .logs.logs import setup_logger

LOGGER = setup_logger()

release = version("topostats")
__version__ = ".".join(release.split("."[:2]))
