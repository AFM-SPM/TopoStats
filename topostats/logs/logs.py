"""
Standardise logging.
"""
import sys
from pathlib import Path
from datetime import datetime
import logging

# pylint: disable=assignment-from-no-return

start = datetime.now()
LOG_INFO_FORMATTER = logging.Formatter(
    fmt="[%(asctime)s] [%(levelname)-8s] [%(name)s] %(message)s", datefmt="%a, %d %b %Y %H:%M:%S"
)
LOG_ERROR_FORMATTER = logging.Formatter(
    fmt="[%(asctime)s] [%(levelname)-8s] [%(name)s] [%(filename)s] [%(lineno)s] %(message)s",
    datefmt="%a, %d %b %Y %H:%M:%S",
)

LOGGER_NAME = "topostats"


def setup_logger(log_name: str = LOGGER_NAME) -> logging.Logger:
    """Setup a standard logger.

    The logger for the module is initialised when the module is loaded (as this functions is called from
    __init__.py). This creates two stream handlers, one for general output and one for errors which are formatted
    differently (there is greater information in the error formatter). To use in modules import the 'LOGGER_NAME' and
    create a logger as shown in the Examples, it will inherit the formatting and direction of messages to the correct
    stream.

    Parameters
    ----------
    log_name : str
        Name under which logging information occurs.

    Returns
    -------
    logging.Logger
        Logger object.

    Examples
    --------

    To use the logger in (sub-)modules have the following.

        import logging
        from topostats.logs.logs import LOGGER_NAME

        LOGGER = logging.getLogger(LOGGER_NAME)

        LOGGER.info('This is a log message.')
    """
    out_stream_handler = logging.StreamHandler(sys.stdout)
    out_stream_handler.setLevel(logging.DEBUG)
    out_stream_handler.setFormatter(LOG_INFO_FORMATTER)

    err_stream_handler = logging.StreamHandler(sys.stderr)
    err_stream_handler.setLevel(logging.ERROR)
    err_stream_handler.setFormatter(LOG_ERROR_FORMATTER)

    file_handler = logging.FileHandler(Path().cwd().stem + f"-{start.strftime('%Y-%m-%d-%H-%M-%S')}.log")
    file_handler.setFormatter(LOG_ERROR_FORMATTER)

    logger = logging.getLogger(log_name)
    logger.setLevel(logging.INFO)
    logger.propagate = True
    if not logger.handlers:
        logger.addHandler(out_stream_handler)
        logger.addHandler(err_stream_handler)
        logger.addHandler(file_handler)

    return logger
