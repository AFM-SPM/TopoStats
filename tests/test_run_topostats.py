"""Test end-to-end running of topostats."""
import logging
from pathlib import Path

import pytest

from topostats.logs.logs import LOGGER_NAME
from topostats.run_topostats import main as run_topostats_main

BASE_DIR = Path.cwd()


@pytest.mark.parametrize("option", ("-h", "--help"))
def test_run_topostats_main_help(capsys, option) -> None:
    """Test the -h/--help flag to run_topostats."""
    try:
        run_topostats_main([option])
    except SystemExit:
        pass
    assert "Process AFM images." in capsys.readouterr().out


def test_run_topostats_process_all(caplog) -> None:
    """Test run_topostats completes without error when no arguments are given."""
    caplog.set_level(logging.INFO)
    # Explicitly force loading of topostats/default_config.yaml as I couldn't work out how to invoke process_all()
    # without any arguments as it defaults to 'sys.argv' as this is wrapped within pytest it picks up the arguments
    # pytest was invoked with (see thread on StackOverflow at https://stackoverflow.com/a/55260580/1444043)
    run_topostats_main(args=["--config", f"{BASE_DIR / 'topostats' / 'default_config.yaml'}"])
    assert "~~~~~~~~~~~~~~~~~~~~ COMPLETE ~~~~~~~~~~~~~~~~~~~~" in caplog.text
    assert "Successfully Processed^1    : 1 (100.0%)" in caplog.text


def test_run_topostats_process_debug(caplog) -> None:
    """Test run_topostats with debugging and check DEBUG messages are logged"""
    # Set the logging level of the topostats logger
    with caplog.at_level(logging.DEBUG, logger=LOGGER_NAME):
        run_topostats_main(args=["--config", f"{BASE_DIR / 'topostats' / 'default_config.yaml'}", "-l", "debug"])
        assert "Configuration after update         :" in caplog.text
        assert "File extension : .spm" in caplog.text
        assert "Images processed : 1" in caplog.text
        assert "~~~~~~~~~~~~~~~~~~~~ COMPLETE ~~~~~~~~~~~~~~~~~~~~" in caplog.text
