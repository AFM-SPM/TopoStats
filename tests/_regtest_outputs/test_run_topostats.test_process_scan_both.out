                   centre_x    centre_y  radius_min  radius_max  radius_mean  radius_median  height_min  height_max  height_median  height_mean       volume        area  area_cartesian_bbox  smallest_bounding_width  smallest_bounding_length  smallest_bounding_area  aspect_ratio threshold  max_feret  min_feret  Contour Lengths  Circular  End to End Distance      Image Name
Molecule Number                                                                                                                                                                                                                                                                                                                                                                       
0                312.321826  135.710289    2.047651    6.802569     4.468766       4.524109   -0.812277   -0.434740      -0.573088    -0.578251   -31.751035   54.908751           106.156919                 5.459192                 13.041404               71.195532      2.388889     lower  13.079421   5.459192        46.737667     False             5.969060  minicircle.spm
0                 18.739027   96.544583    5.914298   20.891859    15.564967      16.245908    0.786263    3.931913       2.103254     2.008786  1280.949667  637.673631          1137.465287                24.428348                 40.849626              997.888853      1.672222     upper  41.334203  24.232548        46.508479      True             0.000000  minicircle.spm
1                400.265969   93.158904    0.790084   21.932873    13.885964      14.786150    0.788546    3.195235       2.171477     2.025761  1278.426351  631.084581          1186.029027                27.450538                 39.092507             1073.110354      1.424107     upper  39.147979  27.450538       115.244383     False            21.470665  minicircle.spm
2                217.791874  173.160884    8.337920   28.460025    17.787064      16.424187    0.788332    3.135538       2.016147     1.904231  1187.324789  623.519376          1592.841865                24.463677                 54.196820             1325.853507      2.215400     upper  54.207678  24.463677        90.207451      True             0.000000  minicircle.spm
3                355.127752  205.914177    3.867763   25.990492    15.545664      16.232298    0.786316    3.117910       2.142770     2.017085  1240.955299  615.222053          1696.558395                17.963499                 49.732773              893.374618      2.768546     upper  49.779222  17.963499        96.013893     False            30.945252  minicircle.spm
4                301.401930  211.610173   10.261029   25.247961    18.007414      18.471047    0.786305    3.007938       1.923541     1.865791  1346.398051  721.623011          1739.509240                32.992930                 44.056166             1453.542026      1.335321     upper  44.124169  32.992930        95.952054      True             0.000000  minicircle.spm
5                432.503984  215.801497    2.896411   23.556389    15.232539      15.827996    0.785852    3.242251       2.201271     2.052962  1260.021782  613.757820          1259.240696                20.726513                 45.916571              951.690429      2.215354     upper  46.207102  20.726513        92.920593     False            22.535415  minicircle.spm
6                194.414424  231.241737    8.973170   23.617666    17.937675      18.161429    0.785752    2.930563       2.074010     1.947770  1339.959985  687.945644          1559.408536                29.666377                 45.552011             1351.363113      1.535476     upper  46.143682  29.666377        97.471758      True             0.000000  minicircle.spm
7                389.730430  274.383831    2.062784   25.547294    16.693765      18.407505    0.790938    3.053797       2.228091     2.054202  1444.761214  703.320094          1770.258141                22.717675                 49.226249             1118.305905      2.166870     upper  49.462005  22.717675        42.431365      True             0.000000  minicircle.spm
8                430.382540  295.946429    2.858291   22.106382    15.491679      16.493713    0.785631    4.416410       2.221225     2.102568  1389.497247  660.857326          1420.794444                26.264609                 42.792910             1123.939040      1.629299     upper  43.023656  26.264609        56.729395      True             0.000000  minicircle.spm
9                122.723943  308.826717    8.901088   27.506537    17.512638      16.620272    0.787003    3.059607       2.265828     2.070719  1430.100586  690.630072          1706.319951                25.553732                 52.521298             1342.115208      2.055328     upper  52.664049  25.553732        75.862966      True             0.000000  minicircle.spm
10                38.889199  382.383429    1.910858   27.319998    17.144151      17.704077    0.786495    2.968241       2.210347     2.039075  1357.987330  665.982143          1234.836806                19.642779                 53.926354             1059.263472      2.745353     upper  54.065681  19.642779        51.707361      True             0.000000  minicircle.spm
11               389.730926  377.575586    9.360143   26.737396    17.552250      16.689506    0.785566    3.050985       2.201184     2.027682  1346.935882  664.273871          1464.233367                27.759526                 47.808537             1327.142337      1.722239     upper  48.080978  27.759526        82.822168      True             0.000000  minicircle.spm
12               261.204440  374.801151    5.782274   25.493809    17.724272      18.612638    0.785335    2.941395       2.101349     1.958538  1359.794642  694.290655          1452.519500                26.526476                 47.827090             1268.684149      1.802994     upper  50.208782  25.683495       100.278899      True             0.000000  minicircle.spm
13               304.678691  391.077914   11.303917   26.520995    18.426317      18.778784    0.785340    3.089508       2.051327     1.930505  1293.220001  669.886766          1665.565455                35.762080                 42.257411             1511.212904      1.181626     upper  48.968624  33.471983        98.671185      True             0.000000  minicircle.spm
14               455.506053  428.067992    4.474904   26.832812    16.790734      17.001231    0.787028    4.001285       2.212375     2.065654  1352.500225  654.756354          1319.762342                21.717876                 51.469374             1117.805484      2.369908     upper  51.646347  21.717876       116.125131     False            16.309580  minicircle.spm
15               220.461913  418.159420    4.149094   27.287016    16.988056      16.254874    0.785245    3.206463       2.128429     1.985213  1378.799064  694.534694          1530.123869                25.498682                 51.345337             1309.238398      2.013647     upper  51.347797  25.498682        91.502865      True             0.000000  minicircle.spm
16               376.110901  428.124688    8.247072   25.064268    18.772413      19.204718    0.786384    3.105426       2.138256     1.983791  1418.477973  715.033961          1530.123869                30.905726                 45.686065             1411.960998      1.478239     upper  48.164652  30.905726       106.408667      True             0.000000  minicircle.spm
17               460.848227  458.689469    8.267206   27.937935    17.001539      16.494286    0.785567    3.150491       2.185607     2.019650  1288.370226  637.917670          1616.513637                21.762554                 52.662810             1146.077257      2.419882     upper  53.178216  21.762554        70.070994      True             0.000000  minicircle.spm
18               419.827308  456.330845   10.613293   24.753580    18.253406      18.889747    0.785293    3.073514       2.169943     2.009988  1389.139268  691.118149          1603.335537                34.580205                 42.978255             1486.196868      1.242857     upper  45.966170  33.806760       101.774488      True             0.000000  minicircle.spm
19                91.004809  468.996912    4.486772   25.295723    16.236281      16.737915    0.785981    3.024085       2.183754     2.029314  1369.315397  674.767543          1600.163031                21.464660                 49.040787             1052.643830      2.284722     upper  49.147706  21.464660       104.448974     False            14.128829  minicircle.spm
20                20.321397  500.754239    2.049918   25.374120    16.105419      16.842532    0.786184    3.083394       2.223461     2.055479  1289.155083  627.179959          1391.021699                20.571420                 50.061003             1029.825933      2.433522     upper  50.157721  20.571420        97.873994     False            20.964590  minicircle.spm
